controllers/front/SmailyCustomerCron.php

```diff
@@ -111,32 +88,117 @@ private function getUserData($customer)
      */
-    private function getUnsubscribers()
+    private function removeUnsubscribers($limit = 1000)
     {
-        $data = array(
-            'list' => 2,
-        );
-        // Api call to Smaily
-        $response = $this->module->callApi('contact', $data);
-        // If successful return unsubscribers.
-        if (isset($response['success'])) {
-            return $response['result'];
-        // If has errors save errors to log.
-        } else {
-            $this->module->logTofile('smaily-cron.txt', $response['error']);
-            return array();
+        $offset = 0;
+        $is_success = true;
+
+        while (true) {
+            // Api call to Smaily
+            $unsubscribers = $this->module->callApi(
+                'contact',
+                array(
+                    'list' => 2,
+                    'limit' => $limit,
+                    'offset' => $offset,
+                )
+            );
+
+            // Stop if error.
+            if (!isset($unsubscribers['success'])) {
+                $is_success = false;
+                break;
+            }
+            // Stop if no more subscribers.
+            if (empty($unsubscribers['result'])) {
+                break;
+            }
+
+            // Remove subscribed status for unsubscribers.
+            $query = 'UPDATE ' . _DB_PREFIX_ . 'customer SET newsletter=0 WHERE email IN (' .
+            implode(
+                ', ',
+                array_map(function ($item) {
+                    return "'" . pSQL($item['email']) . "'";
+                },
+                $unsubscribers['result'])
+            ) . ')';
+            $query_result = Db::getInstance()->execute($query);
+
+            if ($query_result === false) {
+                $is_success = false;
+                break;
+            }
+
+            // Smaily API call offset is considered as page number, not SQL offset!
+            $offset++;
         }
+
+        return $is_success;
+    }
+
+    /**
+     * Send store subscribers data to Smaily.
+     *
+     * @param int $limit subscriber request batch limit.
+     * @return bool Success status.
+     */
+    public function sendSubscribersToSmaily($limit)
+    {
+        $offset = 0;
+        $is_success = true;
+
+        while (true) {
+            $sql = new DbQuery();
+            $sql->select('*');
+            $sql->from('customer', 'c');
+            $sql->where('c.newsletter = 1');
+            $sql->limit(strval($limit), $offset);
+
+            $customers = Db::getInstance()->executeS($sql);
+            // Stop if query fails.
+            if ($customers === false) {
+                $is_success = false;
+                break;
+            }
+            // Stop if no more qustomers.
+            if (empty($customers)) {
+                break;
+            }
+
+            $update_data = array();
+            foreach ($customers as $customer) {
+                $userdata = $this->getUserData($customer);
```

Comment:
Each iteration of `getUserData($customer)` unserializes `SMAILY_SYNCRONIZE_ADDITIONAL` configuration option. 

It would be more optimal to unserialize the option before the while loop and pass unserialized additional fields to `getUserData` as a second argument.

Reply:
Changed the `getUserData($customer, $fields)` function to receive fields from argument and initialize the value only once. https://github.com/sendsmaily/smaily-prestashop-module/pull/10/commits/d96b432a1ae35ba93f4e0827d7c73453d56e113d